src/librustc_typeck/check/cast.rs

```diff
@@ -345,7 +346,7 @@ impl<'tcx> CastCheck<'tcx> {
     {
         // array-ptr-cast.
 
-        if m_expr.mutbl == hir::MutImmutable && m_cast.mutbl == hir::MutImmutable {
+	if !(m_expr.mutbl == hir::MutImmutable && m_cast.mutbl == hir::MutMutable) {
```

Comment:
It should do the opposite of that due to the `!` at the front and only accept casts that are not
`&[T; n]` to `*mut T`. The other 3 casts work with a version of rustc that contains this change.